File: Dev/TemplateFactory/tools/import_azgaar_geojson.js
#!/usr/bin/env node

const path = require("path");
const {
    parseImportArgs,
    readJsonInput,
    renderFrontmatter,
    slugify,
    writeNotes,
    yamlQuote
} = require("./import_map_utils");

const ROOT = process.cwd();
const OUT_DIR = path.join(ROOT, "Mondi", "Luoghi");

function usage() {
    console.log("Uso: npm run import:azgaar -- <file.geojson> [--world \"Nome Mondo\"] [--dry-run]");
}

function getName(feature, index) {
    const props = feature.properties ?? {};
    return props.name || props.Name || props.NAME || props.burg || props.state || props.label || `Luogo Azgaar ${index + 1}`;
}

function inferType(feature) {
    const props = feature.properties ?? {};
    const raw = String(props.type || props.Type || props.kind || props.layer || props.featurecla || "").toLowerCase();
    const geometry = feature.geometry?.type ?? "";

    if (raw.includes("burg") || raw.includes("town") || raw.includes("city") || raw.includes("settlement")) return "insediamento";
    if (raw.includes("state") || raw.includes("country") || raw.includes("province")) return "regione politica";
    if (raw.includes("culture")) return "regione culturale";
    if (raw.includes("religion")) return "regione religiosa";
    if (raw.includes("river")) return "fiume";
    if (raw.includes("route") || raw.includes("road")) return "rotta";
    if (geometry.includes("Polygon")) return "regione";
    if (geometry.includes("Point")) return "luogo";

    return "luogo";
}

function flattenCoords(coords, acc = []) {
    if (!Array.isArray(coords)) return acc;
    if (typeof coords[0] === "number" && typeof coords[1] === "number") {
        acc.push([coords[0], coords[1]]);
        return acc;
    }
    coords.forEach(item => flattenCoords(item, acc));
    return acc;
}

function bbox(feature) {
    const points = flattenCoords(feature.geometry?.coordinates ?? []);
    if (!points.length) return "";

    const xs = points.map(p => p[0]);
    const ys = points.map(p => p[1]);
    return [
        Math.min(...xs),
        Math.min(...ys),
        Math.max(...xs),
        Math.max(...ys)
    ].map(n => Number(n).toFixed(6)).join(", ");
}

function firstPoint(feature) {
    const point = flattenCoords(feature.geometry?.coordinates ?? [])[0];
    return point ? point.map(n => Number(n).toFixed(6)).join(", ") : "";
}

function noteBody({ feature, index, inputName, world }) {
    const props = feature.properties ?? {};
    const name = getName(feature, index);
    const type = inferType(feature);
    const idSource = props.id ?? props.Id ?? props.ID ?? props.i ?? index + 1;
    const coordinate = firstPoint(feature);
    const bounds = bbox(feature);

    const frontmatter = renderFrontmatter("azgaar_import", {
        id: `azgaar-${slugify(name).toLowerCase().replace(/\s+/g, "-")}`,
        nome: yamlQuote(name),
        categoria: 'luogo',
        tipo: yamlQuote(type),
        stato: 'bozza',
        canonico: 'false',
        mondo: world ? `[[${world}]]` : "",
        fonte: 'azgaar',
        file_import: yamlQuote(inputName),
        azgaar_id: yamlQuote(idSource),
        azgaar_layer: yamlQuote(props.layer ?? props.type ?? props.kind ?? ""),
        geometria: yamlQuote(feature.geometry?.type ?? ""),
        coordinate: yamlQuote(coordinate),
        bbox: yamlQuote(bounds),
        fazioni: '[]',
        culture: '[]',
        religioni: '[]',
        personaggi: '[]',
        missioni: '[]',
        segreti: '[]',
        domande_aperte: '[]'
    });

    return `${frontmatter}

# ${name}

> [!luogo] Importato Da Azgaar
> Questa nota e una bozza. Controlla nome, tipo, mondo e collegamenti prima di renderla canonica.

## Uso Nel Mondo

> [!scena] Perche e importante
>

## Dati Importati

\`\`\`json
${JSON.stringify(props, null, 2)}
\`\`\`
`;
}

function main() {
    const args = parseImportArgs(process.argv.slice(2));

    if (!args.input) {
        usage();
        process.exitCode = 1;
        return;
    }

    let input;
    try {
        input = readJsonInput(ROOT, args.input);
    } catch (error) {
        console.error(error.message);
        process.exitCode = 1;
        return;
    }

    const data = input.data;
    if (data.type !== "FeatureCollection" || !Array.isArray(data.features)) {
        console.error("Formato non supportato: serve un GeoJSON FeatureCollection.");
        process.exitCode = 1;
        return;
    }

    const { created, skipped } = writeNotes({
        dryRun: args.dryRun,
        items: data.features,
        outDir: OUT_DIR,
        targetName: getName,
        renderItem: (feature, index) => noteBody({
            feature,
            index,
            inputName: input.inputName,
            world: args.world
        })
    });

    console.log(`${args.dryRun ? "Import simulato" : "Import completato"}: ${created} note, ${skipped} gia esistenti.`);
}

main();

File: Dev/Tools/python/import_azgaar_geojson.py
#!/usr/bin/env python3

from __future__ import annotations

import json
import sys
from pathlib import Path
from typing import Any

from import_map_utils import parse_import_args, read_json_input, render_frontmatter, slugify, write_notes, yaml_quote

sys.dont_write_bytecode = True

ROOT = Path.cwd()
OUT_DIR = ROOT / "Mondi" / "Luoghi"


def usage() -> None:
    print('Uso: npm run import:azgaar -- <file.geojson> [--world "Nome Mondo"] [--dry-run]')


def properties(feature: dict[str, Any]) -> dict[str, Any]:
    props = feature.get("properties")
    return props if isinstance(props, dict) else {}


def geometry(feature: dict[str, Any]) -> dict[str, Any]:
    geom = feature.get("geometry")
    return geom if isinstance(geom, dict) else {}


def get_name(feature: dict[str, Any], index: int) -> str:
    props = properties(feature)
    return str(props.get("name") or props.get("Name") or props.get("NAME") or props.get("burg") or props.get("state") or props.get("label") or f"Luogo Azgaar {index + 1}")


def infer_type(feature: dict[str, Any]) -> str:
    props = properties(feature)
    raw = str(props.get("type") or props.get("Type") or props.get("kind") or props.get("layer") or props.get("featurecla") or "").lower()
    geom_type = str(geometry(feature).get("type") or "")
    if any(marker in raw for marker in ("burg", "town", "city", "settlement")):
        return "insediamento"
    if any(marker in raw for marker in ("state", "country", "province")):
        return "regione politica"
    if "culture" in raw:
        return "regione culturale"
    if "religion" in raw:
        return "regione religiosa"
    if "river" in raw:
        return "fiume"
    if "route" in raw or "road" in raw:
        return "rotta"
    if "Polygon" in geom_type:
        return "regione"
    if "Point" in geom_type:
        return "luogo"
    return "luogo"


def flatten_coords(coords: Any, acc: list[list[float]] | None = None) -> list[list[float]]:
    points = acc if acc is not None else []
    if not isinstance(coords, list):
        return points
    if len(coords) >= 2 and isinstance(coords[0], (int, float)) and isinstance(coords[1], (int, float)):
        points.append([float(coords[0]), float(coords[1])])
        return points
    for item in coords:
        flatten_coords(item, points)
    return points


def bbox(feature: dict[str, Any]) -> str:
    points = flatten_coords(geometry(feature).get("coordinates") or [])
    if not points:
        return ""
    xs = [point[0] for point in points]
    ys = [point[1] for point in points]
    return ", ".join(f"{value:.6f}" for value in [min(xs), min(ys), max(xs), max(ys)])


def first_point(feature: dict[str, Any]) -> str:
    points = flatten_coords(geometry(feature).get("coordinates") or [])
    return ", ".join(f"{value:.6f}" for value in points[0]) if points else ""


def note_body(feature: dict[str, Any], index: int, input_name: str, world: str) -> str:
    props = properties(feature)
    name = get_name(feature, index)
    item_type = infer_type(feature)
    id_source = props.get("id") or props.get("Id") or props.get("ID") or props.get("i") or index + 1
    frontmatter = render_frontmatter(
        {
            "id": f"azgaar-{slugify(name).lower().replace(' ', '-')}",
            "nome": yaml_quote(name),
            "categoria": "luogo",
            "tipo": yaml_quote(item_type),
            "stato": "bozza",
            "canonico": "false",
            "mondo": f"[[{world}]]" if world else "",
            "fonte": "azgaar",
            "file_import": yaml_quote(input_name),
            "azgaar_id": yaml_quote(id_source),
            "azgaar_layer": yaml_quote(props.get("layer") or props.get("type") or props.get("kind") or ""),
            "geometria": yaml_quote(geometry(feature).get("type") or ""),
            "coordinate": yaml_quote(first_point(feature)),
            "bbox": yaml_quote(bbox(feature)),
            "fazioni": "[]",
            "culture": "[]",
            "religioni": "[]",
            "personaggi": "[]",
            "missioni": "[]",
            "segreti": "[]",
            "domande_aperte": "[]",
        }
    )
    return f"""{frontmatter}

# {name}

> [!luogo] Importato Da Azgaar
> Questa nota e una bozza. Controlla nome, tipo, mondo e collegamenti prima di renderla canonica.

## Uso Nel Mondo

> [!scena] Perche e importante
>

## Dati Importati

```json
{json.dumps(props, ensure_ascii=False, indent=2)}
```
"""


def main(argv: list[str] | None = None) -> int:
    args = parse_import_args(argv if argv is not None else sys.argv[1:])
    if not args.input:
        usage()
        return 1
    try:
        source = read_json_input(ROOT, args.input)
    except Exception as error:
        print(error, file=sys.stderr)
        return 1

    data = source["data"]
    if not isinstance(data, dict) or data.get("type") != "FeatureCollection" or not isinstance(data.get("features"), list):
        print("Formato non supportato: serve un GeoJSON FeatureCollection.", file=sys.stderr)
        return 1

    result = write_notes(
        dry_run=args.dry_run,
        items=data["features"],
        out_dir=OUT_DIR,
        target_name=get_name,
        render_item=lambda feature, index: note_body(feature, index, source["input_name"], args.world),
    )
    label = "Import simulato" if args.dry_run else "Import completato"
    print(f"{label}: {result['created']} note, {result['skipped']} gia esistenti.")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
